### PR TITLE
Create storage functions for handling reorgs

### PR DESCRIPTION
**Pull request by iuwqyir**
### TL;DR

Added support for chain reorganization detection and handling in the storage layer.

### What changed?

- Introduced a new `BlockHeader` struct in `block.go`
- Added methods to `ClickHouseConnector` for managing reorg checks and block headers
- Implemented `GetLastReorgCheckedBlockNumber` and `SetLastReorgCheckedBlockNumber` in storage connectors
- Added `LookbackBlockHeaders` and `DeleteDataForBlocks` methods to `IMainStorage` interface
- Updated `MemoryConnector` to support new reorg-related operations
- Added reorg check functionality to `RedisConnector`
- Created a new SQL script for a `cursors` table in ClickHouse
- Modified the staging table creation script to remove the primary key

### CODE REVIEW

**Comment by iuwqyir** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/90?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#91** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/91?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#90** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/90?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#89** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/89?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#88** <a href="https://app.graphite.dev/github/pr/thirdweb-dev/insight/88?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @iuwqyir and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: internal/storage/clickhouse.go [hunk 1]
```diff
@@ -758,6 +758,122 @@ func (c *ClickHouseConnector) GetTraces(qf QueryFilter) (traces []common.Trace,
 	return traces, nil
 }
 
+func (c *ClickHouseConnector) GetLastReorgCheckedBlockNumber(chainId *big.Int) (*big.Int, error) {
+	query := fmt.Sprintf("SELECT cursor_value FROM %s.cursors FINAL WHERE cursor_type = 'reorg'", c.cfg.Database)
+	if chainId.Sign() > 0 {
+		query += fmt.Sprintf(" AND chain_id = %s", chainId.String())
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:764]: does this mean we need an index on chain_id + insert_timestamp?

File: internal/storage/clickhouse.go [hunk 2]
```diff
@@ -758,6 +758,122 @@ func (c *ClickHouseConnector) GetTraces(qf QueryFilter) (traces []common.Trace,
...
+
+	rows, err := c.conn.Query(context.Background(), query)
+	if err != nil {
+		return nil, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var blockHeader common.BlockHeader
+		err := rows.Scan(&blockHeader.Number, &blockHeader.Hash, &blockHeader.ParentHash)
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:799]: We need to start using ScanStruct and just pass the struct here

See: https://clickhouse.com/docs/en/integrations/go#scan-struct

File: internal/storage/clickhouse.go [hunk 3]
```diff
@@ -758,6 +758,122 @@ func (c *ClickHouseConnector) GetTraces(qf QueryFilter) (traces []common.Trace,
...
+	}()
+
+	go func() {
+		defer wg.Done()
+		if err := c.deleteBatch(chainId, blockNumbers, "traces", "chain_id, block_number, is_deleted"); err != nil {
+			saveErrMutex.Lock()
+			saveErr = fmt.Errorf("error deleting traces: %v", err)
+			saveErrMutex.Unlock()
+		}
+	}()
```
Comment (AmineAfia) [inline internal/storage/clickhouse.go:848]: how about we send blocks, logs, transactions and traces in the same batch? whenever an insert fails everything stops and we need a retry

File: internal/storage/memory.go [hunk 4]
```diff
@@ -359,3 +377,44 @@ func (m *MemoryConnector) InsertDataForBlocks(data []common.BlockData) error {
...
+			m.cache.Remove(key)
+		}
+	}
+	return nil
+}
+
+func (m *MemoryConnector) LookbackBlockHeaders(chainId *big.Int, limit int, lookbackStart *big.Int) ([]common.BlockHeader, error) {
+	blockHeaders := []common.BlockHeader{}
+	for _, key := range m.cache.Keys() {
+		if strings.HasPrefix(key, fmt.Sprintf("block:%s:", chainId.String())) {
```
Comment (AmineAfia) [inline internal/storage/memory.go:398]: why hasPrefix and not equal?

File: internal/storage/memory.go [hunk 5]
```diff
@@ -359,3 +377,44 @@ func (m *MemoryConnector) InsertDataForBlocks(data []common.BlockData) error {
...
+			blockNumberStr := strings.Split(key, ":")[2]
```
Comment (AmineAfia) [inline internal/storage/memory.go:399]: wouldn't this be out of range? I think you want 1 not 2

File: internal/tools/clickhouse_create_cursors_table.sql [hunk 6]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE cursors (
+    `chain_id` UInt256,
+    `cursor_type` String,
+    `cursor_value` String,
+    `insert_timestamp` DateTime DEFAULT now(),
+) ENGINE = ReplacingMergeTree(insert_timestamp)
+ORDER BY (chain_id, cursor_type);
```
Comment (AmineAfia) [inline internal/tools/clickhouse_create_cursors_table.sql:7]: this means the data won't be chronological (latest to oldest)?

File: internal/tools/clickhouse_create_staging_table.sql [hunk 7]
```diff
@@ -6,5 +6,5 @@ CREATE TABLE block_data (
     `is_deleted` UInt8 DEFAULT 0,
     INDEX idx_block_number block_number TYPE minmax GRANULARITY 1,
 ) ENGINE = ReplacingMergeTree(insert_timestamp, is_deleted)
-ORDER BY (chain_id, block_number) PRIMARY KEY (chain_id, block_number)
+ORDER BY (chain_id, block_number)
```
[the remark this answers is not in the document]
Comment (iuwqyir) [reply internal/tools/clickhouse_create_staging_table.sql:9]: If no primary key is defined (i.e. PRIMARY KEY was not specified), ClickHouse uses the the sorting key as primary key.
Since these are equivalent, it's not needed

File: internal/storage/clickhouse.go (hunk 1 above)
Comment (iuwqyir) [reply internal/storage/clickhouse.go:764]: the cursors table doesn't need any additional indexes, it is very small. Just 1 entry currently

File: internal/storage/clickhouse.go (hunk 3 above)
Comment (iuwqyir) [reply internal/storage/clickhouse.go:848]: How do you propose sending multiple inserts into **separate tables** in a single batch?

File: internal/storage/clickhouse.go (hunk 2 above)
Comment (iuwqyir) [reply internal/storage/clickhouse.go:799]: Let's have this as an improvement for the future

File: internal/storage/memory.go (hunk 4 above)
Comment (iuwqyir) [reply internal/storage/memory.go:405]: we need items lower than the lookbackStart, so we don't have the block number at that point

File: internal/storage/memory.go (hunk 5 above)
Comment (iuwqyir) [reply internal/storage/memory.go:406]: The format is `type:chainId:blockNumber`

File: internal/tools/clickhouse_create_cursors_table.sql (hunk 6 above)
Comment (iuwqyir) [reply internal/tools/clickhouse_create_cursors_table.sql:7]: It's a single table row that will be overwritten after a successful scan

File: internal/storage/clickhouse.go (hunk 3 above)
Comment (AmineAfia) [reply internal/storage/clickhouse.go:848]: Just googled and found this: https://clickhouse.com/docs/en/cloud/bestpractices/bulk-inserts there is smth called buffer tables that seem to do the job

Comment (iuwqyir) [reply internal/storage/clickhouse.go:848]: It would still be 1 buffer table per table, so it doesn't solve the problem. We could do 1 huge table with columns from each table and figure out if it can buffer for different tables. Or we solve it with materialized views. 
Anyways, this seems like a bigger topic that we should solve when we tackle query atomicity

**Review by AmineAfia — APPROVED** [review]

**Comment by iuwqyir** [comment]
### Merge activity

* **Oct 10, 10:01 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/thirdweb-dev/insight/90).
